fix #1153: use spamVerdict to automatically block spam

groovecoder opened the pull request:
If spam is sent to and thru a Relay alias, MTAs sometimes send bounces
back to Relay with a message like: "Message considered as spam ..."

So, when we receive a bounce for a user and the bounce details includes
the term 'spam', instead of setting a bounce pause on the account,
set a new `auto_block_spam` field.

Then, when we receive any email for a user with `auto_block_spam`, and the
SES `spamVerdict` status is `FAIL`, don’t forward the email.

This should automatically stop us from forwarding any more spam to a
user after their MTA has bounced a first spam message back.

<!-- The following is intended to be helpful to you. Feel free to remove anything that is not. -->

<!-- When fixing a bug: -->

This PR fixes #1153.

## How to test:

Regression testing:  Check that emails are still delivering as expected. May need to do this on dev server.

- [ ] I've added a unit test to test for potential regressions of this bug (or this is a front-end change, where we don't yet have unit test infrastructure).
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

say-yawn reviewed (changes requested):
One change requested and needs migration file for `Profile` changes that adds the `auto_block_spam` field

groovecoder commented on emails/views.py:
```diff
@@ -656,13 +666,19 @@ def _handle_bounce(message_json):
+        if 'spam' in recipient.lower():
+            profile.auto_block_spam = True
+            profile.save()
+            return HttpResponse("OK", status=200)
```
Good catch. Fixed in https://github.com/mozilla/fx-private-relay/pull/1464/commits/6fd0559e5d2bdad132e138933fb5c022ad1cb2e5.

groovecoder commented:
I updated the testing code around this so I could test the new `auto_block_spam` logic during both the bounce-handling and the email-handling. I also ended up refactoring some to make everything more test-able, and I added a few more fixtures we can use to expand on the email-processing testing in the future too.

groovecoder commented on emails/migrations/0038_profile_auto_block_spam.py:
```diff
@@ -0,0 +1,18 @@
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('emails', '0037_reply_add_index_on_created_at'),
```
I'll update this and the migration number + name after we merge https://github.com/mozilla/fx-private-relay/pull/1437.

groovecoder commented on emails/models.py:
```diff
@@ -32,13 +32,18 @@
     }
 
 
-DOMAINS = get_domains_from_settings()
```
Testing this module-level variable was really nasty, so I removed it and replaced all uses with `get_domains_from_settings` instead. I also made `get_domains_from_settings` return mock/test values while in test mode.

say-yawn replied:
I was unsure about using conditionals for tests in feature code but I read some docs talked with @jwhitlock and this is a good approach. When he joins our team I hope we can talk about how we want to structure our test framework or our code so that code is more clear/concise and test conditionals can live in the test code.

groovecoder commented on emails/tests/views_tests.py:
```diff
@@ -1,83 +1,126 @@
-from emails.views import _get_address
+from emails.views import _get_address, _sns_notification
 
-TEST_DOMAINS = {'RELAY_FIREFOX_DOMAIN': 'default.com', 'MOZMAIL_DOMAIN': 'test.com'}
```
Removed this in favor of the new `get_domains_from_settings` function that returns these test values while in test mode.

say-yawn reviewed (changes requested):
LGTM but waiting on merging of #1437 so you can update your migration. Really great to see new tests and glad we are making sure we respect the user's email config and increasing deliverability of emails.

say-yawn commented on emails/views.py:
```diff
@@ -663,6 +673,12 @@ def _handle_bounce(message_json):
+        if (any('spam' in val.lower() for val in recipient.values())):
+            profile.auto_block_spam = True
+            profile.save()
+            continue
```
We record bounce on the user attribute `last_soft_bounce` if it’s user’s email server responding that it is not receiving emails right now (bounced). But when we get a bounce from the email server with the reason being that we are relaying spam, we instead record `auto_block_spam`. This will help us continue to relay emails reliably without being marked as a spammer or putting users relay account on pause.

groovecoder commented:
Merged https://github.com/mozilla/fx-private-relay/pull/1437, rebased, and updated the migration number + name here to depend on that new one.

say-yawn reviewed (approved):
LGTM
